### ISSUE TITLE
Add support for adding templates from a ConfigMap and mounting them as SubPaths

### ISSUE THREAD

**Issue by tsurankov** [hashicorp/vault-k8s#84]
It would be great if vault-injector could mount rendered templates individually to the particular path inside an application container similar to SubPath for ConfigMap volumes.
For example, if an application requires two different large templates to be rendered and placed at different paths:

- application.properties
- server.xml

Supposing templates could be placed inside a config map along with vault agent config, vault-agent would render and place them on a shared volume like it does now. But then rendered configuration files could be mounted to the particular location inside the container using custom annotation for each of the template files:

```
vault.hashicorp.com/agent-inject-template-properties: application.properties
vault.hashicorp.com/agent-inject-template-path-properties: /apps/conf/application.properties
vault.hashicorp.com/agent-inject-template-server: server.xml
vault.hashicorp.com/agent-inject-template-path-server: /usr/local/tomcat/conf/server.xml
```

Otherwise, additional start up scripts are required to place rendered configurations to their proper locations, what is not really handy

**Comment by jasonodonnell** [comment]
Hi @tsurankov, I may be wrong about what you're requesting, but we already support mounting a ConfigMap with a Vault Agent configuration file and supporting templates.  You could choose where to render the templates this way.

See our documentation for an example: https://www.vaultproject.io/docs/platform/k8s/injector/examples/#configmap-example

**Comment by dawidmalina** [comment]
I agree with @tsurankov that it would be desired to have simple way to define (through annotation) destination of the generated template. Keep in mind @jasonodonnell that not all application can be reconfigured to use path like `/vault/secrets/application.properties` and using configmap as you suggested (even if it possible) is not simple way.

**Comment by gunnypatel** [comment]
@jasonodonnell is there currently a way to support using a template file that isn't embedded into the agent configuration file?  Can we use a template that's already in a configmap?

**Comment by virtrixke** [comment]
Same here. I was also expecting the following example to work (in this case for a sql connection setting in the server.xml file in tomcat). Problem is I don't see an option to redirect the original /vault/secrets/server.xml to /usr/local/tomcat/conf in Tomcat using an environment variable or something so I'm a bit stuck:

```
vault-hashicorp-com/secret-volume-path: "/usr/local/tomcat/conf/"
vault.hashicorp.com/agent-inject-template-server.xml: |
          {{- with secret "secret/<namespace>/<tomcatsecrets>" -}}
          <?xml version='1.0' encoding='utf-8'?>
          .......
            <Resource name="jdbc/ProcessEngine"
                        auth="Container"
                        type="javax.sql.DataSource" 
                        factory="org.apache.tomcat.jdbc.pool.DataSourceFactory"
                        uniqueResourceName="process-engine"
                        driverClassName="com.microsoft.sqlserver.jdbc.SQLServerDriver" 
                        url="jdbc:sqlserver://{{ .Data.data.SqlDatabaseString }}"
                        defaultTransactionIsolation="READ_COMMITTED"
                        username="{{ .Data.data.SqlPassword }}"  
                        password="{{ .Data.data.SqlUserName }}"
                        maxActive="20"
                        minIdle="5"
                        maxIdle="20" />
          .......
           {{- end }}
```

**Comment by virtrixke** [comment]
Nevermind... suddenly it started to work. Is there a way to set permissions on a mounted file?

**Comment by virtrixke** [comment]
I use a template to generate the xml files. I did some tests in a container running Camunda BPM (running on tomcat). I get the following result:

![image](https://user-images.githubusercontent.com/26386747/81916930-ce010480-95d4-11ea-82c8-590af8faaf3b.png)

 What about the userid? 100 is not know to Camunda it uses id 1000 (100 = vault uid) so it crashes with 'access denied' on these files. I was expecting uid and gid be like camunda:camunda? Or am I missing something?

**Comment by tsurankov** [comment]
@virtrixke all that I can recommend is adding an init script to the command block of container's spec that will change permissions, ownership and copy the file from the mounted volume to the path that you need for your app.
`cp /vault/secrets/server.xml /usr/local/tomcat/conf/server.xml
chmod 644 /usr/local/tomcat/conf/server.xml
chown uid:gid /usr/local/tomcat/conf/server.xml`
Then start tomcat
Also, It is possible to set permissions and execute a command after secrets rendering in vault-agent's config file. So you can try something like this:
`template {
    source      = "/vault/configs/server.xml.ctmpl"
    destination = "/vault/secrets/server.xml"
    perms = "0644"
    command = "chown uid:gid /vault/secrets/server.xml"
}`

Anyway, I have decided to declare vault-agent init and sidecar containers manually instead of using agent-injector. I find this approach more flexible.

@jasonodonnell The initial issue is that agent-injector doesn't allow mounting rendered secrets separately to different locations using SubPath. Also, I am not sure that it is possible to mount them to a custom folder instead of default one which is /vault/secrets

**Comment by virtrixke** [comment]
Hmm yes. Mounting to SubPath would be a great help and a necessary feature for vault agent injector I suppose.

**Comment by virtrixke** [comment]
In the end we had to do something like this:

```
 command: 
 - /bin/sh
 - -c
 - cp -v /vault/secrets/server.xml /camunda/conf/server.xml && cp -v /vault/secrets/bpm-platform.xml /camunda/conf/bpm-platform.xml && exec /camunda/camunda.sh
```

Its not clean but it will do until SubPath is allowed properly....

**Comment by dcshiman** [comment]
i have noticed that **vault-k8s** agent mutates the deployment and adds
```
 - emptyDir:
      medium: Memory
    name: vault-secrets
```

if the agent allows to override the volume, i think we can achieve this by mounting the volume into the pods.

**Comment by mbwkww** [comment]
Waiting/hoping this feature gets added as well. The functionality we are trying to replicate is the ability to template secrets to arbitrary locations within the file system and the lack of subPaths is preventing that. 

Copying the file from the mounted volume to some other location in the filesystem as suggested above has several permissions issues as well -- namely that if you choose to do this via the template.command hcl block, the user is vault, not root of the final container and the final location of the templated secret (`/etc/somewhere` in the example below) is outside of the mounted, shared volume, as specified via annotation (`/vault/secrets` in the example). 
```
template {
    source      = "/vault/secrets/secret.tpl"
    destination = "/vault/secrets/mysecret"
    command = "cp /vault/secrets/mysecret /etc/somewhere/mysecret"
}
```
We are resorting to symlinks in our final container until functionality that allows the mounting of templated secrets to more locations is added. This could take the form of allowing for subPaths as is suggested above.

**Comment by mrkwtz** [comment]
We would really like this to get implemented. We want to store a file with a secret in the same location of a container where the application files reside (like /opt/app) but when we use the secret-volume-path annotation (with /opt/app) the location gets overwritten with only the secret.

**Comment by norman-zon** [comment]
I too would find it very useful to be able to mount as SubPath.

**Comment by karthick-veluswamy-nbs** [comment]
Using full path for the annotation "vault.hashicorp.com/agent-inject-file-appd" renders with subpath on volumes. 
`vault.hashicorp.com/agent-inject-file-foo: "/app/run/secret.txt"`
